### PR TITLE
Contribution Count Shows Different Numbers  - fixed #443

### LINKED ISSUE #443: Contribution Count Shows Different Numbers
Across different devices and users, the "Contribution Count" on the homepage is showing different numbers.

### PR DESCRIPTION
**Pull request by Ajand**
The root of the bug was the "first" parameter in the GraphQL query which has a default number of 100 and was not handled:
`{
                      grantDonations(where: {lastUpdatedBlockNumber_gte: ${fromBlock}, lastUpdatedBlockNumber_lte: ${blockNumber}}) {
                        grantId
                        tokenIn
                        donationAmount
                        from
                        hash
                        rounds
                        lastUpdatedBlockNumber
                      }
                    }`

so whenever the length of the contributions would go more than 100, the system would act inconsistently.  I created a recursive function that will continue to fetch contributions until there is no more left.

_It only works if the cached data is properly set before "fromBlock" which is not unless there are less than 100 contributions_ **So do the hard refresh and remove the cached data**

### CODE REVIEW

File: app/src/utils/constants.ts
Comment (mds1) [inline app/src/utils/constants.ts:44]: Why 3 and not 2?

File: app/src/utils/data/contributions.ts
Comment (mds1) [inline app/src/utils/data/contributions.ts:59]: Ideally we should have stronger typing than `any`, but we can save that for the PR that handles https://github.com/dcgtc/dgrants/issues/463

File: app/src/utils/constants.ts
Comment (Ajand) [reply app/src/utils/constants.ts:44]: I wasn't sure what should I put, thought maybe some other PR is going to use 2.

**Comment by mds1** [comment]
> I contribute to 6 grants on my own wallet and for other users everything was fine, but the contributor user only fetched 2 contributions. This PR is still an improvement and will fix the problem for most users, but for fixing #443 completely I need to find the root of this problem as well. Maybe we should create a new issue for it.

@Ajand code-wise this looks good, but I'm not sure what exactly you mean here? Which users is this working for and which is it only fetching 2 contributions for?

**Comment by Ajand** [comment]
> > I contribute to 6 grants on my own wallet and for other users everything was fine, but the contributor user only fetched 2 contributions. This PR is still an improvement and will fix the problem for most users, but for fixing #443 completely I need to find the root of this problem as well. Maybe we should create a new issue for it.
> 
> @Ajand code-wise this looks good, but I'm not sure what exactly you mean here? Which users is this working for and which is it only fetching 2 contributions for?

File: app/src/utils/constants.ts
Comment (mds1) [reply app/src/utils/constants.ts:44]: There are no other open PRs changing this, so we should use 2 🙂
```suggestion
  version: 2,
```

**Comment by Ajand** [comment]
> > I contribute to 6 grants on my own wallet and for other users everything was fine, but the contributor user only fetched 2 contributions. This PR is still an improvement and will fix the problem for most users, but for fixing #443 completely I need to find the root of this problem as well. Maybe we should create a new issue for it.
> 
> @Ajand code-wise this looks good, but I'm not sure what exactly you mean here? Which users is this working for and which is it only fetching 2 contributions for?

Let's assume there are 3 users. user 1  will contribute to 10 grants. for user2 and user3, everything is fine, but for user1 we gonna get inconsistent results.

**Comment by mk1r** [comment]
agree with @mds1 comments, so once resolved and approved, I will approve this
good work!

**Comment by Ajand** [comment]
Wrong commit message. There is nothing here for #469. I meant #460.

File: app/src/utils/data/contributions.ts
Comment (Ajand) [reply app/src/utils/data/contributions.ts]: Now we use type ContributionSubgraph instead of any.

File: app/src/utils/constants.ts
Comment (Ajand) [reply app/src/utils/constants.ts]: fixed

**Comment by Ajand** [comment]
There was two problems with #470:
1. The grant donation function would start a race with itself and would override the data it wanted to save so if in a block there would be more than one contribution some of them would not be saved even though they were fetched. I solved that issue by creating an async function call `wait(timeout)` and using a closure called `timoutManager` to put a timeout so the next call would wait so the last call could save its data.

2.  For contributor users for some reason the listener processor would be called 2 times simultaneously so even the in-memory closure wouldn't be able to stop the race. So I added a random timeout in addition to a normal timeout and used a set to prevent processing the same event multiple times.


This approach works fine for now but if we're going to scale ( for example if we have 50 contributions in each block ) it doesn't work. We need to create a queue while processing the events and saving all contributions at once after processing.

Also, it would be nice to find the root of double processing for contributor users so could stop it. Since I'm using a random number there is still a slim chance for some contributors to have some issues.

**Comment by phutchins** [comment]
Whats left here @Ajand ? Is this read for final review? @mk1r thoughts?

**Comment by Ajand** [comment]
Hi @phutchins . 
I think it's finished. It's not the ultimate best solution but it's a pretty good solution for now and I believe it works unless we get to Facebook level scale. I guess we just need @gdixon approval as @mds1 requested.

**Comment by mk1r** [comment]
@phutchins yes looks good to me, are all of @mds1 concerns resolved? @Ajand

**Review by mk1r — APPROVED** [review]
good job!

**Pull request closed without merging**
